### ISSUE TITLE
running MTCL stand-alone

### ISSUE THREAD

**Issue by StefanBloemheuvel** [huankoh/CST-GL#1]
Hi,

Foremost, very interesting work! It looks really cool and congrats with the publication!

I would like to investigate your MTCL module a bit more, and I wondered if it is possible to run it separately somehow.
Or is it that this module is 'embedded' into the whole training procedure?

Kind regards,

Stefan

**Comment by huankoh** [comment]
Hi Stefan,

Thank you for your interest and kind words. We appreciate your engagement with our work.

MTCL is embedded into the whole training procedue as it needs to be train (i.e., optimise for a certain task / case). Here, we allow MTCL to optimize via a forecasting loss in conjunction with the spatiotemporal GNN network when training the entire model for one-step-ahead forecasts. In this manner, CST-GL will find a structure that help it to make accurate forecast. Interestingly, the graph structure that we learned out of the MTCL module helps us in identifying the root cause of anomaly events, suggesting that it indeed learns the underlying structure. 

Nonetheless, it is still possible to run the MTCL module separately - e.g., You can use the MTCL component with some other GNNs (use MTCL.py file in our repo) and optimise together with your new module. 

If your interest leans towards a more generalised graph learner, we also suggest looking into the study "Towards Unsupervised Deep Graph Structure Learning." (https://arxiv.org/abs/2201.06367). It might provide the broader perspective you're searching for in graph learning techniques.

Feel free to ask if you have any other questions or need further clarifications.

Best,
Huan

**Comment by StefanBloemheuvel** [comment]
Thank you very much for your kind reply.

And also for the link to the very interesting paper. I have no clue how that one missed my scope, perhaps because the keywords I used were more in the realm of "Graph Construction" etc. This is exactly what I was looking for!

However, I do have a small question. What did you alter to the learning procedure to simultaneously learn the graph and the target variable(s) (e.g., classes or something else). I have tried that approach before but the neural network never really got it due to not being able to provide any ground truth information about the adjacency matrix. The loss just stayed stable and predictions were horrible. 

Kind regards,

Stefan

**Comment by huankoh** [comment]
My apologies for the very late response. This thread completely slipped through my notifications.

Regarding the "learning procedure to simultaneously learn the graph", we ensured that the value range of the hyper-parameter 𝛼 was higher (e.g., 5-20) and then identified the best one based on the RMSE of the validation set.

On a personal note (not related to this work): It's often non-trivial to arrive at good structures, and it usually requires a lot of trial and error to get things right. I hope you've found ways to solve the learning issue. Please feel free to contact me via email on this.

Best,
Huan
